### PR TITLE
feat: add built-in `jsr` upgrade provider (rework)

### PR DESCRIPTION
**Pull request by JOTSR**
# Summary

Add buit-in support of [jsr](https://jsr.io/) as upgrade provider.
Rework of #687 from main branch (deleting all changes from #679).

# Motivation

Allow to seamlessly upgrade [jsr](https://jsr.io/) hosted cli.

# Changes

- Add `upgrade/provider/jsr.ts`.
- Missing update of [examples/command/upgrade_command.ts](examples/command/upgrade_command.ts) as it relies on non `jsr` package (need discussion/improvement).

### CODE REVIEW

**Comment by JOTSR** [comment]
Cliffy actually assume that the name to the module to install is by default `${name}.ts` in the builtin upgrade command.
```ts
async upgrade(
    { name, from, to, importMap, main = `${name}.ts`, args = [] }:
      UpgradeOptions,
  ): Promise<void>
```
However `jsr` package can use "custom" exports and a cli from this repository can be loaded from `jsr:@scope/name` or `jsr:@scope/name/extensionless_path`. Do you think that the default value for `main: string | undefined` can be:
1. Leave to `undefined`
2. Leave to `${name}`
3. Leave to `${name}.ts` (legacy)

for only `jsr` source in order to avoid supplementary configuration as for others providers ?

eg:
```ts
// deno.json
{
  "name": "@cliffy/cliffy",
  "exports":  {
    ".": "./cliffy.ts",
    "./cliffy": "./cliffy.ts"
}
```
| solution\path |                         default                        |                                exported                                |                                  private                                  |
|:-------------:|:------------------------------------------------------:|:----------------------------------------------------------------------:|:-------------------------------------------------------------------------:|
| 1             | `new JsrProvider({ scope: "cliffy", name: "cliffy" })` | `new JsrProvider({ scope: "cliffy", name: "cliffy", main: "cliffy" })` | `new JsrProvider({ scope: "cliffy", name: "cliffy", main: "cliffy.ts" })` |
| 2             | N.A.                                                   | `new JsrProvider({ scope: "cliffy", name: "cliffy" })`                 | `new JsrProvider({ scope: "cliffy", name: "cliffy", main: "cliffy.ts" })` |
| 3             | N.A.                                                   | `new JsrProvider({ scope: "cliffy", name: "cliffy", main: "cliffy" })` | `new JsrProvider({ scope: "cliffy", name: "cliffy" })`                    |

File: command/upgrade/provider.ts [hunk 1]
```diff
@@ -84,7 +84,12 @@ export abstract class Provider {
       const { latest } = await this.getVersions(name);
       to = latest;
     }
-    const registry: string = new URL(main, this.getRegistryUrl(name, to)).href;
+
+    const registryUrl = this.getRegistryUrl(name, to)
+    const registry: string =
+      registryUrl.startsWith('jsr:') ?
+      `${registryUrl}/${name}` :
```
Comment (c4spar) [inline command/upgrade/provider.ts:91]: Name is already added to registryUrl by getRegistryUrl.

```suggestion
      registryUrl :
```

File: command/upgrade/upgrade_command.ts [hunk 2]
```diff
@@ -55,7 +55,7 @@ export class UpgradeCommand extends Command {
         },
       )
       .option(
-        "--version <version:string:version>",
+        "--version <version:string>",
```
Comment (c4spar) [inline command/upgrade/upgrade_command.ts:58]: You removed the shell completion for the version option 😉

**Review by c4spar — CHANGES_REQUESTED** [review]
Thx for the rework @JOTSR!

File: command/upgrade/provider.ts (hunk 1 above)
Comment (c4spar) [reply command/upgrade/provider.ts:91]: or did you mean `${registryUrl}/${main}`?

Comment (c4spar) [reply command/upgrade/provider.ts:91]: If we use the main option for jsr exports, we need to add a main option to the `Provider` constructor so we can define it differently for different providers. The provider main option would then override the upgrade command main option.

**Comment by JOTSR** [comment]
@c4spar what your though for [provider config](https://github.com/c4spar/deno-cliffy/pull/692#issuecomment-2041625637) defaults ?

**Comment by c4spar** [comment]
I think we can remove the default value for the main option for now. I am currently working on a refactoring of the upgrade command and may move the default value to the provider itself.

**Review by c4spar — APPROVED** [review]
Thx again @JOTSR! LGTM 👍

**Pull request merged**